File: ruby/examples/quickstart.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

#
# remit.md Ruby SDK — Quickstart
#
# All examples use MockRemit, so no API key or network connection is needed.
# To use a real wallet, replace MockRemit.new / mock.wallet with:
#
#   wallet = Remitmd::RemitWallet.new(private_key: ENV["REMITMD_PRIVATE_KEY"])
#

require_relative "../lib/remitmd"

RECIPIENT  = "0x0000000000000000000000000000000000000001"
PAYEE      = "0x0000000000000000000000000000000000000002"
CHALLENGER = "0x0000000000000000000000000000000000000003"

mock   = Remitmd::MockRemit.new
wallet = mock.wallet

puts "=== remit.md Ruby SDK Quickstart ===\n\n"

# 1. Direct payment
puts "1. Direct Payment"
tx = wallet.pay(RECIPIENT, 1.50, memo: "AI inference fee")
puts "   Sent: #{tx.amount} USDC → #{tx.to}"
puts "   Balance after: #{mock.balance} USDC\n\n"

mock.reset

# 2. Escrow
puts "2. Escrow Payment"
escrow = wallet.create_escrow(PAYEE, 50.00, memo: "Code review task")
puts "   Escrow created: #{escrow.id} (#{escrow.status})"
release_tx = wallet.release_escrow(escrow.id)
puts "   Released: #{release_tx.amount} USDC → #{release_tx.to}\n\n"

mock.reset

# 3. Metered Tab
puts "3. Metered Tab (Off-chain billing)"
tab = wallet.create_tab(PAYEE, 10.00)
puts "   Tab opened: #{tab.id} (limit: #{tab.limit} USDC)"
wallet.debit_tab(tab.id, 0.003, "Request #1")
wallet.debit_tab(tab.id, 0.003, "Request #2")
wallet.debit_tab(tab.id, 0.003, "Request #3")
settle_tx = wallet.settle_tab(tab.id)
puts "   Settled: #{settle_tx.amount} USDC on-chain\n\n"

mock.reset

# 4. Payment Stream
puts "4. Payment Stream (Real-time)"
stream = wallet.create_stream(RECIPIENT, 0.001, 10.00)
puts "   Stream started: #{stream.id} at #{stream.rate_per_sec} USDC/sec"
puts "   Deposited: #{stream.deposited} USDC\n\n"

mock.reset

# 5. Bounty
puts "5. Bounty"
bounty = wallet.create_bounty(25.00, "Summarise the top 10 Ethereum EIPs of 2025")
puts "   Bounty posted: #{bounty.id} (#{bounty.award} USDC)"
win_tx = wallet.award_bounty(bounty.id, CHALLENGER)
puts "   Awarded to: #{win_tx.to}\n\n"

mock.reset

# 6. Security Deposit
puts "6. Security Deposit"
dep = wallet.lock_deposit(PAYEE, 100.00, 86_400)
puts "   Deposit locked: #{dep.id} (#{dep.amount} USDC, beneficiary: #{dep.beneficiary})\n\n"

mock.reset

# 7. Reputation check
puts "7. Reputation"
rep = wallet.reputation(RECIPIENT)
puts "   Address: #{rep.address}"
puts "   Score: #{rep.score} | Disputes: #{rep.dispute_rate}\n\n"

puts "All payment models demonstrated successfully."

File: rust/examples/quickstart.rs
//! # remitmd Rust SDK -- Quick Start
//!
//! This example shows the core payment patterns for AI agents.
//! Run with: `cargo run --example quickstart`
//!
//! For tests, use MockRemit (no network required):
//! ```rust
//! let mock = remitmd::MockRemit::new();
//! let wallet = mock.wallet();
//! ```

use remitmd::MockRemit;
use rust_decimal_macros::dec;
use std::time::{SystemTime, UNIX_EPOCH};

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    // ─── Option 1: From environment ───────────────────────────────────────────
    // Set REMITMD_KEY, REMITMD_CHAIN, REMITMD_TESTNET before running.
    //
    // let wallet = Wallet::from_env()?;

    // ─── Option 2: Builder pattern ────────────────────────────────────────────
    // let wallet = Wallet::new("0x<private-key>")
    //     .chain("base")
    //     .testnet()
    //     .build()?;

    // ─── MockRemit for testing (no network, no key) ───────────────────────────
    println!("=== remitmd Rust SDK Quick Start (MockRemit) ===\n");

    let mock = MockRemit::new();
    let wallet = mock.wallet();

    let recipient = "0x0000000000000000000000000000000000000001";
    let payee = "0x0000000000000000000000000000000000000002";

    // 1. Check balance
    let bal = wallet.balance().await?;
    println!("Balance: {} USDC (address: {})", bal.usdc, bal.address);

    // 2. Direct payment -- one-way, instant, no escrow
    let tx = wallet.pay(recipient, dec!(1.50)).await?;
    println!(
        "\n[Direct] Paid {} USDC -> {} (tx: {})",
        tx.amount, tx.to, tx.id
    );

    // Verify with mock assertions
    assert!(mock.was_paid(recipient, dec!(1.50)).await);
    assert_eq!(mock.total_paid_to(recipient).await, dec!(1.50));

    // 3. Escrow -- lock funds, release on delivery
    let escrow = wallet.create_escrow(payee, dec!(100.00)).await?;
    println!(
        "\n[Escrow] Created {} (status: {:?})",
        escrow.id, escrow.status
    );

    // Agent completes work...
    let release_result = wallet.release_escrow(&escrow.id, None).await?;
    println!(
        "[Escrow] Released {} USDC -> {}",
        release_result.amount, release_result.payee
    );

    // 4. Tab -- payment channel for micro-payments
    let tab = wallet.create_tab(payee, dec!(10.00), dec!(0.10)).await?;
    println!(
        "\n[Tab] Opened {} (limit: {} USDC)",
        tab.id, tab.limit_amount
    );

    let closed = wallet.close_tab(&tab.id, 0.10, "0x00").await?;
    println!("[Tab] Closed: status={:?}", closed.status);

    // 5. Stream -- per-second payment flow
    let stream = wallet
        .create_stream(recipient, dec!(0.001), dec!(50.00))
        .await?;
    println!(
        "\n[Stream] Started {} ({} USDC/sec)",
        stream.id, stream.rate_per_second
    );

    // 6. Bounty -- open reward for task completion
    let now = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_secs();
    let bounty = wallet
        .create_bounty(dec!(5.00), "Find cheapest Base RPC endpoint", now + 3600)
        .await?;
    println!(
        "\n[Bounty] Posted {} (amount: {} USDC)",
        bounty.id, bounty.amount
    );

    let awarded = wallet.award_bounty(&bounty.id, 1).await?;
    println!("[Bounty] Awarded: status={:?}", awarded.status);

    // 7. Deposit -- security collateral
    let deposit = wallet
        .lock_deposit(payee, dec!(20.00), now + 86_400)
        .await?;
    println!(
        "\n[Deposit] Locked {} (status: {:?})",
        deposit.id, deposit.status
    );

    // 8. Reputation check
    let rep = wallet.reputation(recipient).await?;
    println!(
        "\n[Reputation] {} score: {}/1000 ({} txs)",
        rep.address, rep.score, rep.transaction_count
    );

    // 9. Spending summary
    let summary = wallet.spending_summary("month").await?;
    println!(
        "\n[Analytics] {} USDC spent this month across {} transactions",
        summary.total_spent, summary.tx_count
    );

    // 10. Budget / operator limits
    let budget = wallet.remaining_budget().await?;
    println!("[Budget] Daily remaining: {} USDC", budget.daily_remaining);

    println!("\nAll patterns demonstrated successfully.");
    Ok(())
}
